File: backend/stock-backend/src/main/java/com/stockapp/entity/PriceHistory.java
package com.stockapp.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "price_history")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PriceHistory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stock_id", nullable = false)
    private Stock stock;

    @Column(nullable = false)
    private LocalDateTime timestamp;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal open;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal high;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal low;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal close;

    @Column(nullable = false)
    private Long volume;

    @Column(nullable = false)
    private String interval;
}

File: ml-service/models/price_history.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    BigInteger,
    DateTime,
    ForeignKey,
    Numeric,
    String,
    UniqueConstraint
)

from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship
)

from models import Base


class PriceHistory(Base):
    __tablename__ = "price_history"

    __table_args__ = (
        UniqueConstraint(
            "stock_id",
            "timestamp",
            name="uk_stock_timestamp"
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)

    stock_id: Mapped[int] = mapped_column(
        ForeignKey("stocks.id"),
        nullable=False
    )

    timestamp: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False
    )

    open: Mapped[Decimal] = mapped_column(
        Numeric(15, 2),
        nullable=False
    )

    high: Mapped[Decimal] = mapped_column(
        Numeric(15, 2),
        nullable=False
    )

    low: Mapped[Decimal] = mapped_column(
        Numeric(15, 2),
        nullable=False
    )

    close: Mapped[Decimal] = mapped_column(
        Numeric(15, 2),
        nullable=False
    )

    volume: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False
    )

    interval: Mapped[str] = mapped_column(
        String(20),
        nullable=False
    )

    stock: Mapped["Stock"] = relationship(
        lazy="joined"
    )